File: src/python/day02.py
def part1(commands):
    horizontal = 0
    depth = 0
    for dir, dist in commands:
        if dir == 'forward':
            horizontal += dist
        elif dir == 'down':
            depth += dist
        elif dir == 'up':
            depth -= dist

    return (horizontal, depth)

def part2(commands):
    horizontal = 0
    depth = 0
    aim = 0
    for dir, delta in commands:
        if dir == 'forward':
            horizontal += delta
            depth += aim * delta
        elif dir == 'down':
            aim += delta
        elif dir == 'up':
            aim -= delta

    return (horizontal, depth)

def main():
    with open('../../data/day02.txt') as f:
        commands = [command.strip().split(' ') for command in f.readlines()]
        commands = [(command[0], int(command[1])) for command in commands]

    horizontal, depth = part1(commands)
    print(horizontal * depth)
    horizontal, depth = part2(commands)
    print(horizontal * depth)


if __name__ == '__main__':
    main()
    

File: src/kotlin/Day02.kt
package day02

import readInput

fun part1(commands: List<List<String>>) : Int {
    var horizontal = 0; var depth = 0
    for (command: List<String> in commands) {
        val distance = command.elementAt(1).toInt()
        when (command.elementAt(0)) {
            "forward" -> horizontal += distance
            "up" -> depth -= distance
            "down" -> depth += distance
        }
    }
    return horizontal * depth
}

fun part2(commands: List<List<String>>) : Int {
    var aim = 0; var horizontal = 0; var depth = 0
    for (command: List<String> in commands) {
        val delta = command.elementAt(1).toInt()
        when (command.elementAt(0)) {
            "forward" -> {
                horizontal += delta
                depth += aim * delta
            }
            "up" -> aim -= delta
            "down" -> aim += delta
        }
    }
    return horizontal * depth
}

fun main() {
    val commands = readInput("day02").map { it.split(' ') }
    println("Part 1: ${ part1(commands) }")
    println("Part 2: ${ part2(commands) }")
}
